runtime/src/qio/auxFilesys/hdfs/qio_plugin_hdfs.c

```diff
@@ -138,6 +138,25 @@ qioerr hdfs_writev(void* fl, const struct iovec* iov, int iovcnt, ssize_t* num_w
   return err_out;
 }
 
+/* 
+   the HDFS3 variable is provided by build system found in files:
+
+   $CHPL_HOME/runtime/etc/Makefile.auxFilesys
+   $CHPL_HOME/runtime/src/qio/auxFilesys/hdfs/Makefile.share
+
+   this slice of code is a simple implementation of an hdfs function 
+   is not currently supported by libhdfs3
+*/
+
+#ifdef HDFS3
+
+tSize hdfsPread(hdfsFS fs, hdfsFile file, tOffset position, void* buffer, tSize length) {
+   hdfsSeek(fs, file, position);
+   return hdfsRead(fs, file, buffer, length);
```

Comment:
This is trickier. The issue is that the code written here is not safe against data races. There is possibly a data race on the position (stored within the file). A Chapel program that opens one file and then 2 channels to that file (in different locations) could - if two calls to this function run at once - return the wrong data in one of the channels.
I'm not sure how to resolve that. Some alternatives that come to mind:
 * don't support hdfs_preadv in the hdfs_function_struct below. I don't recall right now how much other HDFS-related code would break if we did that, though. It might be necessary to adjust some QIO code to emulate pread in this case... so I think it's just putting off the problem.
 * put a lock around these two statements.
 * open up a new copy of the file, seek in the copy, read the data, close the copy

I think I like the last option the best. I will leave it up to you to investigate. I have no idea how bad that would be for performance.